### PR TITLE
Fix InnerSymDCE and move to HW, add support for erasing+setting inner symbols.

### LINKED ISSUE #4391: [InnerSymDCE] inner symbols on ports are not DCE'd, fields not handled/considered
Module ports are not considered for removal in InnerSymDCE, such as this example:

```mlir
firrtl.circuit "Simple" {

  firrtl.module @Simple() {
    %out = firrtl.instance child sym @child @Child(out out : !firrtl.uint<1>)
  }
  firrtl.module @Child(out %out : !firrtl.uint<1> sym @outsym) {
    %c0_ui1 = firrtl.constant 0: !firrtl.uint<1>
    firrtl.strictconnect %out, %c0_ui1 : !firrtl.uint<1>
  }
}
```

Where `@outsym` is dead and can be removed.

### PR DESCRIPTION
**Pull request by dtzSiFive**
Fixes #4391 .

### CODE REVIEW

**Comment by dtzSiFive** [comment]
TODO: Move this pass to HW, where all the InnerSym pieces now reside.

**Comment by dtzSiFive** [comment]
FWIW some notes from working on this:

* non-FIRRTL getPortSymAttr (and now setPortSymAttr) are limited by not wanting to assume the inner symbol attribute is stored under a particular name on a given argument.  We might want to just require it, although hopefully soon we can break out "ports have symbols" while retaining ability to access that via one interface.  Upstream has work on interface inheritance and other goodness that should help quite a lot: https://discourse.llvm.org/t/interface-inheritance-and-dependencies-interface-method-visibility-interface-composition/66380/9 .
* These non-FIRRTL methods only search arguments, but I think we'd want to add them to "results" as well sometimes.
* The various PortInfo-related API don't currently include symbols on ports, re:consistency.
* Much of the common implementation bits, and PortInfo helpers, seem like they probably can be moved to `ModuleImplementation` (`module_like_impl::`) or a new equivalent if needed.

Happily was able to move InnerSymDCE to HW even without tying HW structure ops to IST/IRN yet :+1: .

**Review by mikeurbach — COMMENTED** [review]
Thanks for pushing on this! I only reviewed the new version of the pass using interfaces, and it looks good to me.

File: include/circt/Dialect/HW/HWOpInterfaces.td [hunk 1]
```diff
@@ -37,11 +37,40 @@ def HWModuleLike : OpInterface<"HWModuleLike"> {
     /*methodBody=*/[{}],
     /*defaultImplementation=*/[{ return $_op.getNameAttr(); }]>,
 
+    //===------------------------------------------------------------------===//
+    // Ports and their symbols.
+    //===------------------------------------------------------------------===//
+    // TODO: Refactor this into separate reusable component.
```
[the remark this answers is not in the document]
Comment (dtzSiFive) [reply include/circt/Dialect/HW/HWOpInterfaces.td:43]: Future work.

**Pull request closed without merging**